### PR TITLE
Bug 1217289 - Leak browser iframe through BrowserContextMenu and PinPageSystemDialog r=mhenretty

### PR DESCRIPTION
**Pull request by albertopq**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e98268cd8de5bd8afdd96547838fea668bb0b039

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4e2dfaaea81b9069bffba3415d21840d32cbbfaa

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=59177cf296061265ec308a674ac7795e135e19c7

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9ef989171e7ad75acaa024215afacbd88455e140

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f5bba81b2093183ab53d1e11a1b585a2f620b1d0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3dda85297826618fe3c1229b88b23dba32687712

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cd45859c4d3e44f53ccd461df17e01b757a57dca

File: apps/system/js/pin_page_system_dialog.js [hunk 1]
```diff
@@ -201,10 +202,17 @@
...
+  PinPageSystemDialog.prototype._stop = function() {
+    window.removeEventListener('pin-page-dialog-requestopen', this);
+  };
+
   PinPageSystemDialog.prototype.destroy = function() {
-    this.containerElement.removeChild(this.element);
-    pinDialogInstance = null;
-    this.publish('destroyed');
+    this.stop();
+    SystemDialog.prototype.destroy.apply(this);
```
Comment (mikehenrty) [inline apps/system/js/pin_page_system_dialog.js]: Why can't we just call `this.destroy()`?

File: apps/system/js/app_window.js [hunk 2]
```diff
@@ -926,6 +923,13 @@
         this[componentName] = null;
       }
 
+      for (var propertyName in this.constructor.SUB_MODULES) {
+        if (this[propertyName] && this[propertyName].stop) {
+          this[propertyName].stop();
+        }
+        this[propertyName] = null;
+      }
+
```
Comment (mikehenrty) [inline apps/system/js/app_window.js]: Do we still need to add this stopping logic in `AppWindow` now that we moved `BrowserContextMenu` into `AppWindowManager`? Perhaps this is just for consistency?

Comment (mikehenrty) [inline apps/system/js/app_window.js]: Actually, I'm not sure where we moved BrowserContentMenu in the dependency chain.

Comment (mikehenrty) [inline apps/system/js/app_window.js]: Ah, now I see, it was already a SUB_MODULE, but it's not a SUB_COMPONENT anymore. Ignore the above.

File: apps/system/js/app_window.js [hunk 3]
```diff
@@ -864,13 +863,11 @@
...
-          var moduleName = this.constructor.SUB_MODULES[propertyName];
-          if (moduleName) {
-            this[propertyName] = BaseModule.instantiate(moduleName, this);
-            this[propertyName].start();
-          }
+      for (var propertyName in this.constructor.SUB_MODULES) {
+        var moduleName = this.constructor.SUB_MODULES[propertyName];
+        if (moduleName) {
+          this[propertyName] = BaseModule.instantiate(moduleName, this);
+          this[propertyName].start();
```
Comment (mikehenrty) [inline apps/system/js/app_window.js]: It's a little weird to me that we are instantiating and starting sub_modules in the installSubComponents function (shouldn't this be handled when the app_window is created?), but I guess it's already been like this so it's not part of this bug.

File: apps/system/js/pin_page_system_dialog.js (hunk 1 above)
Comment (mikehenrty) [inline apps/system/js/pin_page_system_dialog.js]: Derp, you are calling the parent. nm

File: apps/system/js/system_dialog_manager.js [hunk 4]
```diff
@@ -55,9 +55,12 @@
...
      */
     configs: {
-      listens: ['system-dialog-created',
-                'simlockcreated',
+      listens: ['system-dialog-started',
+                'system-dialog-stopped',
+                'simlockstarted',
+                'simlockstopped',
                 'actionmenucreated',
+                'actionmenudestroyed',
```
Comment (mikehenrty) [inline apps/system/js/system_dialog_manager.js]: When does `actionmenudestroyed` get called. I only see activities.js and multi_screen_contoller.js use it, and they don't declare it in their dependency lists but rather instantiate it directly.

**Pull request merged**